### PR TITLE
Added Metadata to Offerings

### PR DESCRIPTION
**Pull request by varyamereon**
Hope I've done this okay, this is actually my first ever pull request!!🫣

I've added a Dictionary of Metadata to OfferingDto. The platform implementations required converting the native objects to dotnet objects, I've tested as much as possible on both platforms and the conversions seem to work well.

### CODE REVIEW

**Comment by varyamereon** [comment]
Resolves #32

**Comment by Kebechet** [comment]
Hello @varyamereon, 
big thanks and also congratulation to your 1st PR 🎉🎉

In following 2 days I will look at your changes.

**Comment by varyamereon** [comment]
Thanks! Hope you like it or maybe you see a better way of doing it. Either way I find the metadata pretty useful.

**Comment by Kebechet** [comment]
@varyamereon, I've been reviewing your PR, specifically the parsing functionality for the metadata object. First off, I want to say thanks for the work you've put into this.

I'm curious about the decision to parse the metadata object into concrete types. According to the [documentation](https://www.revenuecat.com/docs/tools/offering-metadata) you shared, the metadata is a JSON object. On platforms like `Android` and `iOS`, this means dealing with `IDictionary<string, Java.Lang.Object>` and `NSDictionary<NSString, NSObject>`, respectively.

I wonder if we could explore the benefits of maintaining the metadata values as strings within our .NET integration. My thinking is that since consumers of the library will likely need to implement some form of parsing logic on their end anyway, we might simplify the process by treating these values as strings. Here's a rough idea of how this might look:

`OfferingDto`:
```cs
public Dictionary<string, string?> Metadata { get; init; } = new();
```
and parsing to string like:
``` cs
//Android
Metadata = offer.Metadata.ToDictionary(item => item.Key, item => item.Value?.ToString())

//iOS
Metadata = offer.Metadata.ToDictionary<KeyValuePair<NSString, NSObject>, string, string?>(item => item.Key,
    item => item.Value?.ToString()),
```

I'd love to hear your thoughts on this. Perhaps there are aspects I haven't considered so please feel free to share your perspective.

**Comment by varyamereon** [comment]
At first I thought it would be possible to actually just return the JSON data either as a string or a JSON object that the user could parse into a class themselves. But it doesn't look like that's possible, so then I thought it would make way more sense to return native types, I'm not sure I want to be parsing every string it returns into bools or numbers every time! I'm not sure how that would even look with arrays or dictionaries... The type of information you can use as metadata with RC is fairly simple so I thought it might be a good idea to return the dotnet types as that's sort of what I'd expect from a user perspective. What do you think?

**Comment by Kebechet** [comment]
> At first I thought it would be possible to actually just return the JSON data either as a string or a JSON object that the user could parse into a class themselves. But it doesn't look like that's possible, 
Yes, RevenueCat flattens 1st level of the JSON structure and it provides it as a dictionary (for some strange reason)

> so then I thought it would make way more sense to return native types, I'm not sure I want to be parsing every string it returns into bools or numbers every time! I'm not sure how that would even look with arrays or dictionaries... The type of information you can use as metadata with RC is fairly simple so I thought it might be a good idea to return the dotnet types as that's sort of what I'd expect from a user perspective. What do you think?

But still even in your solution you have to parse it on client side because you are returning `object` right ? And regarding parsing complex types like dictionaries. Simple `JsonSerializer.Deserialize` would do the job wouldnt it ?

**Comment by varyamereon** [comment]
What do you think about just reserializing the Dictionary back into a JSON string? That would give the user complete control, if they wanted to deserialize it into a custom class they could, otherwise they could just do whatever they wanted with it. Haven't tested it but something like:

```
            var offeringDto = new OfferingDto()
            {
                Identifier = offer.Identifier,
                AvailablePackages = offer.AvailablePackages.ToPackageDtoList(),
                IsCurrent = offer.Identifier == offerings.Current?.Identifier,
                Metadata = JsonSerializer.Serialize(offer.Metadata)
            };
```

**Comment by Kebechet** [comment]
This seems like the best solution, really smart. It would be 1:1 to the JSON we insert into RevenueCat so I am all for this one.

But the code you posted with `JsonSerialized` is crashing on Android (did not test iOS). So probably each platform must have it's custom json serialization.
On android I was able to do it like:
```
Metadata = offer.Metadata.IsNullOrEmpty()
    ? null
    : new JSONObject((IDictionary)offer.Metadata).ToString().ToString()
```
1st `ToString` is into `Java` object and second one is into `.NET` `string`.

So please do something similar on iOS (or try the JsonSerializer) and we will use this approach (returning the whole JSON as one string).

**Comment by varyamereon** [comment]
Glad you figured it out on Android as I was struggling.... The iOS implementation is done too. In both cases a nullable string is returned.

File: src/Maui.RevenueCat.InAppBilling/Platforms/iOS/Extensions/RCOfferingsExtensions.cs [hunk 1]
```diff
@@ -12,72 +12,27 @@ public static List<OfferingDto> ToOfferingDtoList(this RCOfferings offerings)
...
             var offeringDto = new OfferingDto()
             {
                 Identifier = offer.Identifier,
                 AvailablePackages = offer.AvailablePackages.ToPackageDtoList(),
                 IsCurrent = offer.Identifier == offerings.Current?.Identifier,
-                Metadata = offer.Metadata.ToDictionary<KeyValuePair<NSString, NSObject>, string, object?>(item => item.Key,
-                    item => item.Value.ToDotNetObject())
+                Metadata = jsonString
             };
             offeringDtos.Add(offeringDto);
```
Comment (Kebechet) [inline src/Maui.RevenueCat.InAppBilling/Platforms/iOS/Extensions/RCOfferingsExtensions.cs:33]: Here we are mixing 2 concerns:
1) mapping native object to `OfferingDto`
2) a way how to parse iOS `NSDictionary` to json string

So I would propose to extract the serialization logic to separate method in `NsDictionaryExtensions`

Something like:
```
    internal static string? ToJson<T, U>(this NSDictionary<T, U> dictionary)
        where T : NSObject
        where U : NSObject
    {
        if (dictionary.IsNullOrEmpty())
        {
            return null;
        }

        var jsonData = NSJsonSerialization.Serialize(dictionary, NSJsonWritingOptions.PrettyPrinted, out var error);

        return error is null
            ? NSString.FromData(jsonData, NSStringEncoding.UTF8)
            : null;
    }
```

and then simply use:
```
Metadata = offer.Metadata.ToJson()
```

but after implementing this into iOS we are creating small inconsistency between Android and iOS code. Good practice is to keep them as similar as possible. So also please create `IDictonaryExtensions` in Android platform extensions folder where would be similar implementation

```
using Maui.RevenueCat.InAppBilling.Extensions;
using Org.Json;
using System.Collections;

namespace Maui.RevenueCat.InAppBilling.Platforms.Android.Extensions;

internal static class IDictonaryExtensions
{
    internal static string? ToJson<T, U>(this IDictionary<T, U> dictionary)
    {
        if (dictionary.IsNullOrEmpty())
        {
            return null;
        }

        return new JSONObject((IDictionary)dictionary).ToString().ToString();
    }
}
```

That way mapping in both implementations is identical and the only thing that is different are underlying type operations extracted into extension methods.

**Review by Kebechet — CHANGES_REQUESTED** [review]
I just added a final touch to the solution. Please take a look.

**Comment by varyamereon** [comment]
Good point, have now added those extension methods.

**Review by Kebechet — APPROVED** [review]

**Comment by Kebechet** [comment]
> Good point, have now added those extension methods.

Perfect, approved. Thank you for the cooperation. I am going to release a new version

**Comment by varyamereon** [comment]
Cool, thanks for your help!

**Pull request merged**